File: src/main/java/com/anfema/ionclient/caching/index/PageCacheIndex.java
package com.anfema.ionclient.caching.index;

import android.content.Context;

import com.anfema.ionclient.IonConfig;
import com.anfema.ionclient.pages.IonPageUrls;
import com.anfema.ionclient.pages.models.Page;

import org.joda.time.DateTime;

public class PageCacheIndex extends CacheIndex
{
	/**
	 * Last changed info received from server in the page response
	 */
	private DateTime lastChanged;

	public PageCacheIndex( DateTime lastChanged )
	{
		this.lastChanged = lastChanged;
	}

	public DateTime getLastChanged()
	{
		return lastChanged;
	}

	public void setLastChanged( DateTime lastChanged )
	{
		this.lastChanged = lastChanged;
	}

	public boolean isOutdated( DateTime serverDate )
	{
		return lastChanged.isBefore( serverDate );
	}

	// save & retrieve

	public static PageCacheIndex retrieve( String requestUrl, IonConfig config, Context context )
	{
		return CacheIndexStore.retrieve( requestUrl, PageCacheIndex.class, config, context );
	}

	public static void save( Page page, IonConfig config, Context context )
	{
		save( page.identifier, page.last_changed, config, context );
	}

	public static void save( String pageIdentifier, DateTime pageLastChanged, IonConfig config, Context context )
	{
		String url = IonPageUrls.getPageUrl( config, pageIdentifier );
		PageCacheIndex cacheIndex = new PageCacheIndex( pageLastChanged );
		CacheIndexStore.save( url, cacheIndex, config, context );
	}
}

File: src/main/java/com/anfema/ionclient/caching/index/PageCacheIndex.kt
package com.anfema.ionclient.caching.index

import android.content.Context
import com.anfema.ionclient.CollectionProperties
import com.anfema.ionclient.pages.IonPageUrls
import com.anfema.ionclient.pages.models.Page
import org.joda.time.DateTime

class PageCacheIndex(
    /**
     * Last changed info received from server in the page response
     */
    private val lastChanged: DateTime,
) : CacheIndex() {

    fun isOutdated(serverDate: DateTime?): Boolean =
        lastChanged.isBefore(serverDate)

    companion object {

        @JvmStatic
        fun retrieve(
            requestUrl: String,
            collectionProperties: CollectionProperties,
            context: Context,
        ): PageCacheIndex? =
            CacheIndexStore.retrieve(requestUrl, collectionProperties, context)

        @JvmStatic
        fun save(page: Page, collectionProperties: CollectionProperties, context: Context) {
            save(page.identifier, page.last_changed, collectionProperties, context)
        }

        @JvmStatic
        fun save(
            pageIdentifier: String,
            pageLastChanged: DateTime,
            collectionProperties: CollectionProperties,
            context: Context,
        ) {
            val url = IonPageUrls.getPageUrl(collectionProperties, pageIdentifier)
            val cacheIndex = PageCacheIndex(pageLastChanged)
            CacheIndexStore.save(url, cacheIndex, collectionProperties, context)
        }
    }
}
